460_Codyjwhite_MethodforPermContext

Codyjwhite opened the pull request:
Adds a new method to create ongoing context allowing permanent context and ongoing dynamic context to be included in HTTP requests.

Updated LLM Classes to include tokenCount, contextSize and maxOutputTokens attributes.
tokenCount:
•	tokenCount is set at 0.25. As a standard most LLMs average 1 token = 4 characters
•	tokenCount is used to estimate token size of prompts.

_______________________________________________________________________________________________________________________________________
contextSize:
•	contextSize is now an optional attribute when constructing an LLM object defaulting to 4000 token. 
•	contextSize being optional means it can be left out of the constructor in existing features. 
•	New features will be able to set a context size when creating an LLM object. The default of 4000 tokens is enough for existing features. (Note: This new parameter is only relevant to the new pathway for context building.) 
•	4000 is chosen as a safe option form most models. Many models support much larger (Such as Deepseek and ChatGPT being 128k) unless a feature needs this much context it is often a net negative to fill the context.
_______________________________________________________________________________________________________________________________________
maxOutputTokens
•	maxOutputTokens is now an optional attribute when constructing an LLM object defaulting to 1000 tokens
•	maxOutputTokens being optional means it can be left out of the constructor in existing features. 
•	maxOutputTokens is used to calculate a token budget for features that need to police context size.

Note: Base class for the LLM was also updated.
_______________________________________________________________________________________________________________________________________
New Data Class:
ChatLog.dart add implementation for the PermTokens class, and the chatTurn class
PerTokens:
•	Class to hold permanent context.
•	Core is an optional attribute that handles permanent context. This can be as simple as the system message, or as complex as personality instructions. It does have a default prompt, though new features should implement new instructions here.
•	Modules is an empty list that will hold optional context. For the Essay builder this will be Modes, and helper instructions, for the interactive personas this will hold various bot instructions.
_______________________________________________________________________________________________________________________________________
ChatTurn:
•	Class that represents a single chat turn.
•	Object holds the role (user, assistant, system)
•	Object holds the content. (the user message or the systems output.
•	Class is responsible for creating of Objects -> Json and Json -> Objects
_______________________________________________________________________________________________________________________________________
New service files (Token_Utils.dart and ContextBuilder.dart):
Token_Utils.dart:
•	Class used to give a rough but safe estimate of token size of a prompt.
_______________________________________________________________________________________________________________________________________
ContextBuilder.dart:
Contains helper method buildContext() to assemble and properly order the prompt inside a List<Map<String, Dynamic>>.
Basic logic:
1.	Set token budget
2.	Put core at the start
3.	Iterate through modules adding prompts checking token budget exiting if budget is reached.
4.	Puts most recent user prompt input
5.	Iterates through chat log adding messages in reverse order until all messages are added or context limit is reached inserting them in order before the most recent prompt.
6.	Returns the list.
_______________________________________________________________________________________________________________________________________

Added new Method for chat completion.

- chat() *Currently only using Deepseek and OpenAI will work on Grok and perplexity later*

- This new method takes in either a <Map<String, Dynamic>> (such as a return from ContextBuilder) or a single String prompt.

- chat() allows more freedom in enabling the LLM through optional parameters such as Temperature. *See inline comments for more details*

- Existing feature should be able to call this method with relative ease. Though this new new method does use named parameters ( chat(prompt : userPrompt, temperature : 0.65) )

Codyjwhite committed:
Adds a new pathway to create ongoing context allowing permanent context and ongoing dynamic context to be included in HTTP requests to the LLM

rappleb1 commented:
As a potential test case, can you update the existing chat interface to use these new methods before moving on to the new features?

Codyjwhite committed:
Made some small changes to types, added as test file. Ran formatting commands

Codyjwhite committed:
Formatting fixes

Codyjwhite committed:
Think I know why it's failing.

Added LLM classes. This includes a new chat completion method that was gonna be a seperate PR. 🤞

Codyjwhite committed:
Trying again.

Codyjwhite committed:
Fixed Typo

Codyjwhite committed:
Changed dafault max token size from 500 to 1000

rappleb1 committed:
Merge remote-tracking branch 'origin/team_e' into 460_Codyjwhite_MethodForPermContext

rappleb1 reviewed (approved):
I think this looks good - tests pass.
